src/accessibility/describe.js

```diff
@@ -0,0 +1,304 @@
+const fds = '_fds'; //Fallback description
+const ft = '_ft'; //Fallback Table
+const fte = '_fte_'; //Fallback Table Element
+const lb = '_Label'; //Label container
+const lds = '_lds'; //Label description
+const lt = '_lt'; //Label Table
+const lte = '_lte_'; //Label Table Element
+
+/**
+ * Creates a screen-reader accessible description for the canvas.
+ * The first parameter should be a string with a description of the canvas.
+ * The second parameter is optional. If specified, it determines how the
+ * description is displayed.
+ *
+ * <code class="language-javascript">describe(text, LABEL)</code> displays
+ * the description to all users as a <a
+ * href="https://en.wikipedia.org/wiki/Museum_label" target="_blank">
+ * tombstone or exhibit label/caption</a> in a
+ * <code class="language-javascript">&lt;div class="p5Label"&gt;&lt;/div&gt;</code>
+ * adjacent to the canvas. You can style it as you wish in your CSS.
+ *
+ * <code class="language-javascript">describe(text, FALLBACK)</code> makes the
+ * description accessible to screen-reader users only, in
+ * <a href="https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Hit_regions_and_accessibility" target="_blank">
+ * a sub DOM inside the canvas element</a>. If a second parameter is not
+ * specified, by default, the description will only be available to
+ * screen-reader users.
+ *
+ * @method describe
+ * @param  {String} text      description of the canvas
+ * @param  {Constant} [display] either LABEL or FALLBACK (Optional)
+ *
+ * @example
+ * <div>
+ * <code>
+ * describe('pink square with red heart in the bottom right corner', LABEL);
+ * background('pink');
+ * fill('red');
+ * noStroke();
+ * ellipse(67, 67, 20, 20);
+ * ellipse(83, 67, 20, 20);
+ * triangle(91, 73, 75, 95, 59, 73);
+ * </code>
+ * </div>
+ *
+ * <div>
+ * <code>
+ * let x = 0;
+ * function draw() {
+ *   if (x > 100) {
+ *     x = 0;
+ *   }
+ *   background(220);
+ *   fill(0, 255, 0);
+ *   ellipse(x, 50, 40, 40);
+ *   x = x + 0.1;
+ *   describe('green circle at x pos ' + round(x) + ' moving to the right');
+ * }
+ * </code>
+ * </div>
+ *
+ */
+
+p5.prototype.describe = function(t, d) {
+  const cnvId = this.canvas.id;
+  p5._validateParameters('describe', arguments);
+  //Creates a sub DOM inside of the canvas element and populates
+  //it with description text.
+  if (typeof t !== 'string') {
+    return;
+  }
+  t = this._descriptionText(t);
+  if (document.getElementById(cnvId + ds) === null) {
+    document.getElementById(cnvId).innerHTML =
+      '<div id="' +
+      cnvId +
+      ds +
+      '" role="region" aria-label="Canvas Description"><p id="' +
+      cnvId +
+      fds +
+      '"></p></div>';
+  } else if (document.getElementById(cnvId + fds) === null) {
+    document
+      .getElementById(cnvId + ft)
+      .insertAdjacentHTML('beforebegin', '<p id="' + cnvId + fds + '"></p>');
+  }
+  if (document.getElementById(cnvId + fds).innerHTML !== t) {
```

Comment:
i'm not sure this test is necessary. could you just set the `innerHTML`? or is there a performance advantage to not reseting if it's already set?

Reply:
If an html element is constantly being reset some screenreaders will read the html element as empty and will not be able to access its content. This why we are reseting only when the content is different.

Reply:
ah! interesting, makes sense. thanks for explaining